elements/lisk-p2p/src/peer/outbound.ts

```diff
@@ -0,0 +1,339 @@
+		}
+		this._socket.connect();
+	}
+
+	public disconnect(code: number = 1000, reason?: string): void {
+		if (this._socket) {
+			this._socket.destroy(code, reason);
+			this._unbindHandlersFromOutboundSocket(this._socket);
+		}
+	}
+
+	// All event handlers for the outbound socket should be bound in this method.
+	private _bindHandlersToOutboundSocket(outboundSocket: SCClientSocket): void {
+		outboundSocket.on('error', (error: Error) => {
+			this.emit(EVENT_OUTBOUND_SOCKET_ERROR, error);
+		});
+
+		outboundSocket.on('connect', () => {
+			this.emit(EVENT_CONNECT_OUTBOUND, this._peerInfo);
+		});
+
+		outboundSocket.on('connectAbort', () => {
+			this.emit(EVENT_CONNECT_ABORT_OUTBOUND, this._peerInfo);
+		});
+
+		outboundSocket.on('close', (code: number, reason: string) => {
+			this.emit(EVENT_CLOSE_OUTBOUND, {
+				peerInfo: this._peerInfo,
+				code,
+				reason,
+			});
+		});
+
+		// Bind RPC and remote event handlers
+		outboundSocket.on(REMOTE_EVENT_RPC_REQUEST, this._handleRawRPC);
+		outboundSocket.on(REMOTE_EVENT_MESSAGE, this._handleRawMessage);
+		outboundSocket.on('postBlock', this._handleRawLegacyMessagePostBlock);
+		outboundSocket.on(
+			'postSignatures',
+			this._handleRawLegacyMessagePostSignatures,
+		);
+		outboundSocket.on(
+			'postTransactions',
+			this._handleRawLegacyMessagePostTransactions,
+		);
+	}
+
+	// All event handlers for the outbound socket should be unbound in this method.
+	/* tslint:disable-next-line:prefer-function-over-method*/
+	private _unbindHandlersFromOutboundSocket(
+		outboundSocket: SCClientSocket,
+	): void {
+		// Do not unbind the error handler because error could still throw after disconnect.
+		// We don't want to have uncaught errors.
+		outboundSocket.off('connect');
+		outboundSocket.off('connectAbort');
+		outboundSocket.off('close');
+
+		// Unbind RPC and remote event handlers
+		outboundSocket.off(REMOTE_EVENT_RPC_REQUEST, this._handleRawRPC);
+		outboundSocket.off(REMOTE_EVENT_MESSAGE, this._handleRawMessage);
+		outboundSocket.off('postBlock', this._handleRawLegacyMessagePostBlock);
+		outboundSocket.off(
+			'postSignatures',
+			this._handleRawLegacyMessagePostSignatures,
+		);
+		outboundSocket.off(
+			'postTransactions',
+			this._handleRawLegacyMessagePostTransactions,
+		);
+	}
+}
+
+export const connectAndRequest = async (
+	basicPeerInfo: P2PPeerInfo,
+	procedure: string,
+	nodeInfo?: P2PNodeInfo,
+	peerConfig?: PeerConfig,
+): Promise<ConnectAndFetchResponse> =>
+	new Promise<ConnectAndFetchResponse>(
+		(resolve, reject): void => {
+			const legacyNodeInfo = nodeInfo
+				? convertNodeInfoToLegacyFormat(nodeInfo)
+				: undefined;
+			// Add a new field discovery to tell the receiving side that the connection will be short lived
+			const requestPacket = {
+				procedure,
+			};
+			// Ideally, we should JSON-serialize the whole NodeInfo object but this cannot be done for compatibility reasons, so instead we put it inside an options property.
```

Comment:
do we need to do this for all the options?
I think it was originally something like,
```
const clientOptions = {
...basicPeerInfo,
...peerConfig,
query: querystring.stringify({
					...legacyNodeInfo,
					options: JSON.stringify(legacyNodeInfo),
				}),
}
```